File: solutions/4036-sum-of-weighted-modes-in-subarrays/sum-of-weighted-modes-in-subarrays.cpp
class Solution {
public:
    long long modeWeight(vector<int>& nums, int k) {
        long long res = 0, n = nums.size();
        unordered_map<int, int> freq; // {num, frequency}
        map<int, set<int>> vals;      // Frequencies and their nums sorted
        for (int i = 0; i < n; i++) {
            // 1. Remove prev number that is not included in current window.
            if (i - k >= 0) {
                int pprevFreq = freq[nums[i - k]], pprevVal = nums[i - k];
                vals[pprevFreq].erase(pprevVal);
                if (vals[pprevFreq].empty()) {
                    vals.erase(pprevFreq);
                }
                freq[pprevVal]--;
                if (freq[pprevVal] > 0) {
                    vals[freq[pprevVal]].insert(pprevVal);
                }
            }
            
            // 2. Remove frequency for current number, it will be updated later.
            int prevFreq = freq.count(nums[i]) > 0 ? freq[nums[i]] : 0;
            if (prevFreq > 0) {
                vals[prevFreq].erase(nums[i]);
                if (vals[prevFreq].empty())
                    vals.erase(prevFreq);
            }
            
            // 3. Make updated with current number.
            freq[nums[i]]++;
            vals[freq[nums[i]]].insert(nums[i]);
            if (i - k + 1 >= 0) {
                int maxFreq = vals.rbegin()->first;
                int mode = *vals[maxFreq].begin();
                res += 1LL * maxFreq * mode;
            }
        }
        return res;
    }
};

File: solutions/4036-sum-of-weighted-modes-in-subarrays/sum-of-weighted-modes-in-subarrays.java
class Solution {
    // time = O(nlogn), space = O(n)
    public long modeWeight(int[] nums, int k) {
        int n = nums.length;
        long res = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        TreeSet<int[]> set = new TreeSet<>((o1, o2) -> o1[0] != o2[0] ? o2[0] - o1[0] : o1[1] - o2[1]);
        for (int i = 0, j = 0; i < n; i++) {
            if (i - j + 1 > k) {
                set.remove(new int[]{map.get(nums[j]), nums[j]});
                map.put(nums[j], map.get(nums[j]) - 1);
                if (map.get(nums[j]) > 0) set.add(new int[]{map.get(nums[j]), nums[j]});
                j++;
            }
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            set.add(new int[]{map.get(nums[i]), nums[i]});
            if (i - j + 1 == k) {
                int[] x = set.first();
                res += 1L * x[0] * x[1];
            }
        }
        return res;
    }
} 